### DISCUSSION TITLE
想实现一个自动更新sqlite数据库的功能

### DISCUSSION THREAD

**Discussion by sid033** [dotnetcore/FreeSql#1189] [Q&A]
我使用sqlite作为数据库，项目中有一个需求：需要在项目运行的状态下，通过U盘插入时自动搜索更新数据库文件。因目标机器可能不会有屏幕，这样就需要这样一个机制，U盘自动更新。问题是：我插入U盘后，替换现有文件的时候将会因为文件被freesql占用，而无法替换，这样就需要手动关闭freesql的连接，清空连接池。我尝试调用dispose()，是可以替换的，但是在恢复的时候，我会收到 一个例外：The ObjectPool has been disposed。请问如何关闭数据库，再重新打开呢？

**Comment by 2881099** [comment]
fsql.Dispose() 之后，需要重新 new FreeSqlBuilder().Build() 创建

**Reply by sid033** [reply]
我在项目中替换完数据库文件后，重新调用了build创建的。

**Reply by sid033** [reply]
public static void Rebuild()
        {
            try
            {
                string connectStr = common.Config.GetStringValue("SqliteConn");
                if (string.IsNullOrEmpty(connectStr))
                {
                    connectStr = "Data Source=data/config.db;Version=3;Pooling=False;Max Pool Size=100;";
                }
                    _fsql = new FreeSql.FreeSqlBuilder()
                        .UseConnectionString(FreeSql.DataType.Sqlite, connectStr)
                        .UseAutoSyncStructure(false)
                        .Build();
            }
            catch (Exception e)
            {
                Log.Log.Error("无法连接sqlite数据库：" + e.ToString());
            }
        }

**Reply by sid033** [reply]
DataCollector.Jobs.autoUpgrade.AfterUpgrade += new Action(()=> {
                DataCollector.Db.Database.Rebuild();
                startServer(true);
            });

**Comment by 2881099** [comment]
那就是你的代码有些地方还在用老的 IFreeSql 对象

**Reply by sid033** [reply]
感谢2881099！首先，我就一个入口使用了这个freesql对象，关闭服务时，将关闭这个唯一入口，所以不是其他的地方使用了这个对象。而是freesql内部的一些机制造成。后来我用别的方法解决了。问题解决方式：使用UseConnectionFactory替代UseConnectionString。估计后者使用了连接池一些技术，造成相同的连接串，使用了旧的连接。但是这个UseConnectionFactory却不是推荐的方式，不知道有没有后来者愿意在这个问题深入的讨论一下，给我这样一个小白多些教诲。
